export/mlab_export.py

```diff
@@ -0,0 +1,591 @@
+
+
+class Error(Exception):
+  """Base error type for this file."""
+
+
+class TimeOptionError(Error):
+  """An error related to export times or ranges."""
+
+
+def init_global():
+  global HOSTNAME
+  # NOTE: This should be the hostname of root context, not slice context.
+  HOSTNAME = socket.gethostname()
+  logging.basicConfig(format='%(message)s', level=logging.INFO)
+
+
+class LastTime(object):
+  """Manages interactions with export timestamp."""
+
+  def __init__(self, filename):
+    self._filename = filename
+    self._lock_handle = None
+    self._first_creation = not os.path.exists(self._filename)
+
+  def __enter__(self):
+    self.open()
+    return self
+
+  def __exit__(self, *unused_args):
+    self.close()
+
+  def open(self):
+    """Opens the timestamp file, and creates the file if it does not exist."""
+    if self._first_creation:
+      # Since the file does not exist, use mode 'w' to create it.
+      self._lock_handle = open(self._filename, 'w')
+    else:
+      # File *does* exist, so use mode 'r' to preserve the mtime on the file.
+      self._lock_handle = open(self._filename, 'r')
+
+  def close(self):
+    """Closes the timestamp file."""
+    if self._lock_handle is not None:
+      self._lock_handle.close()
+
+  def get_mtime(self):
+    """Returns the file mtime, or zero if file was just created."""
+    if self._first_creation:
+      # Indicate that there is no timestamp, so the caller can use a default.
+      return 0
+    return int(os.stat(self._filename).st_mtime)
+
+  def update_mtime(self, mtime):
+    """Updates the atime & mtime on the export timestamp file."""
+    os.utime(self._filename, (mtime, mtime))
+
+  def lock(self):
+    """Attempt to acquire advisory lock on export timestamp file.
+
+    This method returns immediatly.
+
+    Returns:
+      bool, True if the lock is acquired, False otherwise.
+    """
+    # Attempt an exclusive lock without blocking. On success, fcntl.flock()
+    # returns no value. On error, it throws IOError EAGAIN "temporarily
+    # unavailable."
+    try:
+      fcntl.flock(self._lock_handle, fcntl.LOCK_EX|fcntl.LOCK_NB)
+      return True
+    except IOError:
+      return False
+
+
+def align_timestamp(timestamp, step):
+  """Adjusts 'timestamp' down to a multiple of 'step' size.
+
+  Args:
+    timestamp: int, timestamp in seconds since the epoch.
+    step: int, interval to align.
+  Returns:
+    int, timestamp adjusted to multiple of step.
+  """
+  return timestamp - (timestamp % step)
+
+
+def default_end_time(options):
+  """Calculates a default, end timestamp.
+
+  Args:
+    options: flags.FlagValues, the runtime options. These values are read:
+        options.length, options.step, options.update, options.ts_offset.
+  Returns:
+    int, timestamp in seconds since the epoch.
+  """
+  ts_current = int(time.time())
+
+  # Align current ts to a multiple of length.
+  ts_end = align_timestamp(ts_current, options.length)
```

Comment:
Great. All of your time-related comments encouraged me to revisit what was going on. Two things: I was going out of my way to make it possible for a human to specify all the parameters whereas only two should be required (e.g. start + length (one of your comments)); also, I was beginning with `ts_end` and then working backward to calculate `ts_start`. This is counter-intuitive. It turns out that working from ts_start (which is usually ts_previous from the last run) and then calculating ts_end made things simpler too.

How does it look now? 99b050b86b4fdbcd816db86a265165755ac400cd